flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManagerDriver.java

```diff
@@ -684,6 +694,7 @@ public void onGetContainerStatusError(ContainerId containerId, Throwable throwab
 
         @Override
         public void onStopContainerError(ContainerId containerId, Throwable throwable) {
+            trackerOfReleasedResources.arriveAndDeregister();
```

Comment:
What if a container errors that is not released? Wouldn't this thwart our bookkeeping?

Reply:
Not sure that I follow. If we start the container release process we basically count up (register). When the container release process completes successfully we count down (deregister). The same happens, when the release process fails (onStopContainerError). If we start a new release process, we would still count up at the start and count down at the end.

Are you saying that scenarios exist where we count up, but never count down or vice versa? This can only happen, if the callback is for some reason never called. Possibly due to a failure in the YARN clients. Even in this (unlikely?) case, the containers will be killed shortly afterwards anyway. (That is, if the shutdown procedure is initated over YARN's kill application command.)